### DISCUSSION TITLE
Configuration examples

### DISCUSSION THREAD

**Discussion by spookylukey** [python-lsp/python-lsp-server#176] [Configuration]
Could you expand on the [configuration](https://github.com/python-lsp/python-lsp-server#configuration) section with the help of examples?

For example:
- how do I actually "Change the pylsp.configurationSources setting to ['flake8'] " as suggested?
- then, how do you actually set  do `"pylsp.plugins.pydocstyle.enabled": true` - on a per workspace basis, or globally?

Other people are also struggling with configuration e.g. https://github.com/python-lsp/python-lsp-server/issues/162

Thanks!

**Comment by ccordoba12** [comment]
As I said several times before, this is not something we can help users with here. It depends on your editor or IDE, i.e. Neovim, Emacs, Sublime, Jupyterlab-LSP, etc.

Most of them allow to set options in json format, so they are sent to the server when it starts. But how to do that, where to put those options and in what format is editor/IDE specific. And most of us here don't use any of the options I mentioned above, so we have no clue about all that.

So the appropriate place to ask for help is in your editor/IDE issue tracker or forum.

**Comment by spookylukey** [comment]
The [current docs say](https://github.com/python-lsp/python-lsp-server#configuration):

> configuration is loaded from zero or more configuration sources. Currently implemented are:
> 
> * pycodestyle: discovered in ~/.config/pycodestyle, setup.cfg, tox.ini and pycodestyle.cfg.
> * flake8: discovered in ~/.config/flake8, setup.cfg, tox.ini and flake8.cfg
> 
> The default configuration source is pycodestyle. Change the pylsp.configurationSources setting to ['flake8'] in order to respect flake8 configuration instead.
>
> Overall configuration is computed first from user configuration (in home directory), overridden by configuration passed in by the language client, and then overridden by configuration discovered in the workspace.
>
> To enable pydocstyle for linting docstrings add the following setting in your LSP configuration: "pylsp.plugins.pydocstyle.enabled": true

This sounds like 3 config sources:
1. user config
2. "configuration passed in by the language client" i.e. sent from editor
3. configuration discovered in the workspace

You are talking about number 2 here I think. I'm asking about 1 and 3. Doesn't this text mean that you can put something in the workspace to affect configuration - a e.g. a `setup.cfg` file in a folder? If not, what does it mean?

**Comment by ccordoba12** [comment]
> I'm asking about 1 and 3

Nop, you're asking about 2 as well because neither pycodestyle, flake8 nor any other plugin can be enabled/disabled from user config files.

And this applies to issue #162 too.

> Doesn't this text mean that you can put something in the workspace to affect configuration - a e.g. a setup.cfg file in a folder?

Sure, if the plugin is enabled. After that, you can configure it as you would in a regular project.

**Comment by carltongibson** [comment]
@ccordoba12 Are you able to point to **any** examples of adjusting the configuration in a real-project? 

I appreciate that all editors do it differently, but the language server protocol a constant, and so we should be able to extrapolate. 

At the moment I  (and it looks like others) am struggling to get any of [the 3rd Party Plugins (under the python-lsp org) listed in the README](https://github.com/python-lsp/python-lsp-server#3rd-party-plugins) to register... 

> Sure, if the plugin is enabled.

I'm guessing that's my issue, but I can't get it working, and I can't find a single example to start working from. 

Thanks! 🎁

**Comment by spookylukey** [comment]
Thanks so much for your reply @ccordoba12 I think I'm getting there, but I'm still quite confused.

If I understand rightly (which I'm not at all confident of), I would rewrite the Configuration section like this:

> # Configuration
>
> Like all language servers, configuration can be passed from the client (i.e. the editor/IDE). The details of how this is done depend on the editor or plugin that you are using to communicate with python-lsp-server. The configuration options available are documented in CONFIGURATION.md
> 
> python-lsp-server depends on other tools, like flake8 and pycodestyle. These tools can be configured via settings passed from the client (first paragraph above), or alternatively other configuration sources. The following sources are available:
>
> - pycodestyle: discovered in `~/.config/pycodestyle`, `setup.cfg`, `tox.ini` and `pycodestyle.cfg`.
> - flake8: discovered in `~/.config/flake8`, `setup.cfg`, `tox.ini` and `flake8.cfg`
>
> The default configuration source is pycodestyle. Change the `pylsp.configurationSources` setting (in the value passed in from your client) to `['flake8']` in order to respect flake8 configuration instead.
>
> The configuration options available in these config files (`setup.cfg` etc) are documented in the relevant tools:
> - [flake8 configuration](https://flake8.pycqa.org/en/latest/user/configuration.html)
> - [pycodestyle configuration](https://pycodestyle.pycqa.org/en/latest/intro.html#configuration)
>
> Overall configuration is computed first from user configuration (in home directory), overridden by configuration passed in by the language client, and then overridden by configuration discovered in the workspace.
>
> As an example, to change the list of errors that pycodestyle will ignore, assuming you are using the `pycodestyle` configuration source (the default), you can:
> 
> 1. Add the following to your ~/.config/pycodestyle:
> 
> [pycodestyle]
> ignore = E226,E302,E41
>
> 2. Set the `pylsp.plugins.pycodestyle.ignore` config value from your editor
> 3. Same as 1, but added to `setup.cfg` file in the root of the project.

This is an attempt to make sense of this, I'm still quite fuzzy on some details, perhaps it's an improvement?

**Comment by ccordoba12** [comment]
> Are you able to point to any examples of adjusting the configuration in a real-project?

See [this comment](https://github.com/python-lsp/python-lsp-server/issues/171#issuecomment-1069026008) for JupyterLab-LSP, although I think the `serverSettings` key is specific to that project.

> This is an attempt to make sense of this, I'm still quite fuzzy on some details, perhaps it's an improvement?

Yes, it is! Thanks for your suggestions @spookylukey!

Could you submit a pull request with them? Thanks!

**Comment by antoineFrau** [comment]
Hello,

Thanks for opening this issue on this subject, I have been facing the same issue about configuring the language server. In my case, I'm using the client ([monaco-languageclient](https://github.com/TypeFox/monaco-languageclient/issues/201)). I understand that the problem might be on the client-side, but looking into the websocket event, I can't find the workspace/didChangeConfiguration being called or even set by default in the capabilities of the server (I tried to manually do that but it's not working either). Client capabilities are set correctly too, with the [didChangeConfiguration dynamicRegistration](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeConfiguration) set to true.

So my question is the following, is the responsibility of the language server to ask for the configuration to the client, or is it the opposite, the client is in charge of sending it during the initialization?
In my case using monaco-languageclient, the didChangeConfiguration functions are never called. I guess I will open an issue on the lib to ask about it.

Another question to make it clearer for the documentation what should be the format of the configuration JSON (client side)?

```
{
    "pylsp": {
        "configurationSources": ["flake8"],
        "plugins": {
            "flake8": {
                "enabled": true,
                "exclude": [],
                "hangClosing": false,
                "ignore": ["F401", "E133", "E203", "W503"],
                "maxLineLength": 10,
                "indentSize": 245,
                "perFileIgnores": [],
                "select": []
            }
        }
    }
}
```
Or
```
{
    "pylsp.configurationSources": ["flake8"],
    "pylsp.plugins.flake8.enabled": true,
    "pylsp.plugins.flake8.exclude": [],
    "pylsp.plugins.flake8.hangClosing": false,
    "pylsp.plugins.flake8.ignore": ["F401", "E133", "E203", "W503"],
    "pylsp.plugins.flake8.maxLineLength": 10,
    "pylsp.plugins.flake8.indentSize": 245,
    "pylsp.plugins.flake8.perFileIgnores": [],
    "pylsp.plugins.flake8.select": []
}
```

**Comment by ccordoba12** [comment]
I'm going to move this to our newly opened Discussions because it's not really a bug but a petition for help.

I'll post further comments there.

**Comment by ccordoba12** [comment]
> I can't find the workspace/didChangeConfiguration being called or even set by default in the capabilities of the server

That request needs to be sent from the client to the server. Our response is handled here:

https://github.com/python-lsp/python-lsp-server/blob/581da5d11f5e90879388994bdc26d6d36de7a642/pylsp/python_lsp.py#L386-L392

> So my question is the following, is the responsibility of the language server to ask for the configuration to the client

No

> or is it the opposite, the client is in charge of sending it during the initialization?

This is the case. The client needs to send that request to the server.

In Spyder that's the second request we send after we receive a response for the `initialize` one. In other words, as soon as we detect that the server is up and running, and responded to our `initialize` request, then we send a `workspace/didChangeConfiguration` with the necessary parameters to configure it.

> Another question to make it clearer for the documentation what should be the format of the configuration JSON (client side)?

I think your first option is the right one, but I'm not really sure. Please experiment with that and let us know which one works for you.

**Comment by dholth** [comment]
I was trying to configure with https://github.com/superlou/lapce-python's plugin.toml

```
[configuration.options]
# Options passed to the LSP server
"pylsp.plugins.autopep8.enabled" = false
"pylsp.plugins.yapf.enabled" = false
"pylsp.plugins.mccabe.enabled" = false
```

These are included as part of the initialization `'initializationOptions': {'pylsp.plugins.autopep8.enabled': False, 'pylsp.plugins.mccabe.enabled': False, 'pylsp.plugins.yapf.enabled': False}, ...` but don't seem to have an effect.

**Reply by dholth** [reply]
This passs the expected `{'pylsp':{'pylsp.plugins....}` nested structure but doesn't appear to do anything either, reading through the -vv verbose logs...
```
[configuration.options.pylsp]
# Options passed to the LSP server
"pylsp.plugins.autopep8.enabled" = false
"pylsp.plugins.yapf.enabled" = false
"pylsp.plugins.mccabe.enabled" = false
```

**Reply by rchl** [reply]
Configuration should come through `workspace/didChangeConfiguration`, not `initializationOptions`.

Also, the structure should be like:

```
{
  "settings": {
    "pylsp": {
      "configurationSources": [
        "pycodestyle"
      ],
      "plugins": {
        "autopep8": {
          "enabled": false
        },
...
```

There should be unexpanded keys like `pylsp.plugins`. It should be `pylsp: { plugins: {} }`

**Reply by dholth** [reply]
https://github.com/python-lsp/python-lsp-server/blob/a9d503c5a766188628f554f53bde1116a5b58bc2/pylsp/python_lsp.py#L287 the initializationOptions dict is merged into `Config()`

Maybe could print `Config()` when running under vscode to compare.

**Reply by rchl** [reply]
In that case the issue is likely with the configuration object being un-expanded (`{'pylsp':{'pylsp.plugins....}`).

**Reply by dholth** [reply]
In vscode (not supported by python-lsp-server) I set `"python.trace.server": "verbose",` and checked the Output: Python Language Server window to see how the protocol works. It does turn dotted keys from `.vscode/setings.json` into nested structures, and sends something like this

```
[Trace - 4:57:30 PM] Sending notification 'workspace/didChangeConfiguration'.
Params: {
    "settings": {
        "python": {
            "autoComplete": {
                "extraPaths": []
            },
            "envFile": "${workspaceFolder}/.env",
...
            "terminal": {
                "activateEnvInCurrentTerminal": false,
                "activateEnvironment": true,
                "executeInFileDir": false,
                "focusAfterLaunch": false,
                "launchArgs": []
            },
...
            "testing": {
                "autoTestDiscoverOnSaveEnabled": true,
...
            },
            "venvFolders": [],
            "venvPath": "",
...
            "trace": {
                "server": "verbose"
            },
            "configurationSources": [
                "spam",
                "ham",
                "jam",
                "spinach"
            ],
            "plugins": {
                "autopep8": {
                    "enabled": true
                },
                "flake8": {
                    "config": null,
                    "enabled": "true",
                    "exclude": "shucks",
                    "executable": "maybe",
                    "filename": "yes",
                    "hangClosing": true,
                    "ignore": true
                }
            }
        }
    }
}
```

for settings including

```
    "python.configurationSources": [
        "spam",
        "ham",
        "jam",
        "spinach",
    ],
    "python.plugins.autopep8.enabled": true,
    "python.plugins.flake8.config": null,
    "python.plugins.flake8.enabled": "true"
    "python.plugins.flake8.exclude": "shucks",
    "python.plugins.flake8.executable": "maybe",
    "python.plugins.flake8.filename": "yes",
    "python.plugins.flake8.hangClosing": true,
    "python.plugins.flake8.ignore": true
```

Those options should all start with "pylsp." instead of "python." to be sent to python-lsp-server; it only sends the language server's own configuration and not other keys from `settings.json`.

The editor I'm using has a toml config file, and the toml format expands dotted keys into nested objects in the same way.

```
[configuration.options.settings]
# Options passed to the LSP server
plugins.autopep8.enabled = false
plugins.yapf.enabled = false
```

**Reply by rchl** [reply]
Options passed to this server should be nested and not include dots. Like:

```js
{
  "settings": {
    "pylsp": {
      "configurationSources": [
        "pycodestyle"
      ],
      "plugins": {
        "autopep8": {
          "enabled": false
        },
      },
      "rope": {
        "extensionModules": null,
        "ropeFolder": null
      }
    }
  }
}
```

> The editor I'm using has a toml config file, and the toml format expands dotted keys into nested objects in the same way.

Does it...? In https://github.com/python-lsp/python-lsp-server/discussions/176#discussioncomment-3271746 you seem to be saying that it doesn't.

**Reply by dholth** [reply]
You have to leave the dotted toml keys un-quoted. If you quote them, then a literal "." appears in the key. If they are unquoted, it makes a nested object.

**Comment by doolio** [comment]
[An example for Emacs+Eglot](https://github.com/python-lsp/python-lsp-server/discussions/293).
